pkg/api/customization/vsphere/listers.go

```diff
@@ -0,0 +1,401 @@
+	}
+
+	return data, nil
+}
+
+func listVirtualMachines(ctx context.Context, finder *find.Finder, t string) ([]string, error) {
+	if t == "" {
+		t = "both"
+	}
+	vms, err := finder.VirtualMachineList(ctx, "*")
+	if err != nil {
+		return nil, err
+	}
+
+	var data []string
+	for _, vm := range vms {
+		switch t {
+		case "both":
+			data = append(data, vm.InventoryPath)
+		case "templates":
+			if isTemplate, err := vm.IsTemplate(ctx); err == nil && isTemplate {
+				data = append(data, vm.InventoryPath)
+			}
+		case "virtual-machines":
+			if isTemplate, err := vm.IsTemplate(ctx); err == nil && !isTemplate {
+				data = append(data, vm.InventoryPath)
+			}
+		}
+	}
+
+	return data, nil
+}
+
+func getNetworkInventoryPath(nr object.NetworkReference) string {
+	var r string
+	switch nr.Reference().Type {
+	case "Network":
+		r = nr.(*object.Network).InventoryPath
+	case "OpaqueNetwork":
+		r = nr.(*object.OpaqueNetwork).InventoryPath
+	case "DistributedVirtualPortgroup":
+		r = nr.(*object.DistributedVirtualPortgroup).InventoryPath
+	case "DistributedVirtualSwitch":
+		r = nr.(*object.DistributedVirtualSwitch).InventoryPath
+	case "VmwareDistributedVirtualSwitch":
+		r = nr.(*object.VmwareDistributedVirtualSwitch).InventoryPath
+	}
+
+	return r
+}
+
+func getSoapFinder(ctx context.Context, cc *corev1.Secret, dc string) (*find.Finder, error) {
+	c, err := getSoapClient(ctx, cc)
+	if err != nil {
+		return nil, err
+	}
+
+	finder := find.NewFinder(c.Client, true)
+
+	if dc != "" {
+		dataCenter, err := finder.Datacenter(ctx, dc)
+		if err != nil {
+			return nil, err
+		}
+
+		finder.SetDatacenter(dataCenter)
+	}
+
+	return finder, nil
+}
+
+func getSoapClient(ctx context.Context, cc *corev1.Secret) (*govmomi.Client, error) {
+	u, err := url.Parse(fmt.Sprintf("https://%s:%s/sdk", cc.Data[dataFields["host"]], cc.Data[dataFields["port"]]))
+	if err != nil {
+		return nil, err
+	}
+
+	u.User = url.UserPassword(string(cc.Data[dataFields["username"]]), string(cc.Data[dataFields["password"]]))
+	c, err := govmomi.NewClient(ctx, u, true)
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func getContentLibraryManager(ctx context.Context, cc *corev1.Secret) (*library.Manager, error) {
+	soap, err := getSoapClient(ctx, cc)
+	if err != nil {
+		return nil, err
+	}
+
+	mgr := library.NewManager(rest.NewClient(soap.Client))
+	ui := url.UserPassword(string(cc.Data[dataFields["username"]]), string(cc.Data[dataFields["password"]]))
+	err = mgr.Login(ctx, ui)
+	if err != nil {
+		return nil, err
+	}
+
+	return mgr, nil
+}
+
+func getTagsManager(ctx context.Context, cc *corev1.Secret) (*tags.Manager, error) {
+	soap, err := getSoapClient(ctx, cc)
+	if err != nil {
+		return nil, err
+	}
+
+	mgr := tags.NewManager(rest.NewClient(soap.Client))
+	ui := url.UserPassword(string(cc.Data[dataFields["username"]]), string(cc.Data[dataFields["password"]]))
+	err = mgr.Login(ctx, ui)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Feels like this ^ block could be abstracted so it's not repeated.

Reply:
they're different types, probably not worth the headache with how govmomi sets up a tags.Manager and a tags.Library. im ok to look harder but it's only called twice.